**Ex3.4 Mathematica Modelling of given problem**  
Let us denote following symbols as decision variables for the corresponding Drone Models  
Sankalp:x0  
Kartav: x1  
Saahas: x2  
Vijay: x3  
Lakshya: x4

Objective is to maximise total carrying capacity(kgs)  
Max. 15*x0 + 10*x1 + 20*x2 + 10*x3 + 15*x4

Constraints  
Budget constraint: 4*x0+ 3.5*x1+ 5*x2+ 4*x3+ 4.5*x4 <=29  
Space constraint: 3*x0+ 3.25*x1+ 4*x2+ 2.75*x3+ 2.5*x4 <=37

x0,x1,x2,x3,x4 are all non negative integers.

In [29]:
!pip install -q pyomo
from pyomo.environ import * 
import numpy as np
import pandas as pd
!apt-get install -y -qq coinor-cbc

In [30]:
data_csvfile = pd.read_csv('lab4_ex3.csv')

model=ConcreteModel()

N = len(data_csvfile.index)
M = 2

col_indices = range(N)

model.x = Var(col_indices,domain=NonNegativeIntegers)

model.objective = Objective(expr=summation(data_csvfile['Capacity'], model.x),sense=maximize)

model.constraints = ConstraintList()

#budget constraint
model.constraints.add(summation(data_csvfile['Cost'], model.x) <= 29)

#space constraint
model.constraints.add(summation(data_csvfile['Space Required'], model.x) <= 37)

model.pprint()


2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    2 : {1, 2}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  None :  None : False :  True : NonNegativeIntegers
          1 :     0 :  None :  None : False :  True : NonNegativeIntegers
          2 :     0 :  None :  None : False :  True : NonNegativeIntegers
          3 :     0 :  None :  None : False :  True : NonNegativeIntegers
          4 :     0 :  None :  None : False :  True : NonNegativeIntegers

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 15*x[0] + 10*x[1] + 20*x[2] + 10*x[3] + 

In [32]:
opt_cbc=SolverFactory('cbc')
result = opt_cbc.solve(model)

print('\nMaximum Capacity = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print('Number of Drone',data_csvfile['Drone'][i], model.x[i].value)



Maximum Capacity =  115.0

Decision Variables
Number of Drone Sankalp 1.0
Number of Drone Kartav 0.0
Number of Drone Saahas 5.0
Number of Drone Vijay 0.0
Number of Drone Lakshya 0.0


**3.6** Maximum Capacity =  115.0

Decision Variables  
Number of Drone Sankalp 1.0  
Number of Drone Kartav 0.0  
Number of Drone Saahas 5.0  
Number of Drone Vijay 0.0  
Number of Drone Lakshya 0.0   

In [33]:
# 3.7 removing integer constraints
model.x.domain=NonNegativeReals

opt_cbc=SolverFactory('cbc')
result = opt_cbc.solve(model)

print('\nMaximum Capacity = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print('Number of Drone',data_csvfile['Drone'][i], model.x[i].value)


Maximum Capacity =  116.0

Decision Variables
Number of Drone Sankalp 0.0
Number of Drone Kartav 0.0
Number of Drone Saahas 5.8
Number of Drone Vijay 0.0
Number of Drone Lakshya 0.0


**3.8** Solution of MILP can not be obtained by just rounding off the solution of LP for this problem as the solution obtained in two situations assigns diffrent values of decison variables and hence the objective funtion.

In [34]:
#3.9 with LP
#changing constraints
model.constraints[1].deactivate()
model.constraints[2].deactivate()

#budget constraint
model.constraints.add(summation(data_csvfile['Cost'], model.x) <= 41)

#space constraint
model.constraints.add(summation(data_csvfile['Space Required'], model.x) <= 47)



opt_cbc=SolverFactory('cbc')
result = opt_cbc.solve(model)

print('\nMaximum Capacity = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print('Number of Drone',data_csvfile['Drone'][i], model.x[i].value)




Maximum Capacity =  164.0

Decision Variables
Number of Drone Sankalp 0.0
Number of Drone Kartav 0.0
Number of Drone Saahas 8.2
Number of Drone Vijay 0.0
Number of Drone Lakshya 0.0


In [35]:
#3.9 with MILP

model.x.domain=NonNegativeIntegers

opt_cbc=SolverFactory('cbc')
result = opt_cbc.solve(model)

print('\nMaximum Capacity = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print('Number of Drone',data_csvfile['Drone'][i], model.x[i].value)


Maximum Capacity =  160.0

Decision Variables
Number of Drone Sankalp 0.0
Number of Drone Kartav 0.0
Number of Drone Saahas 8.0
Number of Drone Vijay 0.0
Number of Drone Lakshya 0.0


**3.9** On increasing the budget to 41 crores and space avaiable to 47 sq.m. we see the follwin change in objective function  
LP objective value increased from 116 to 164
MILP objective value increased from 115 to 160

In [37]:
# 3.10 (a)LP

model.x.domain=NonNegativeReals

model.constraints[3].deactivate()
model.constraints[4].deactivate()

#cost constraint
model.constraints.add(summation(data_csvfile['Cost'], model.x) <= 39)

#space constraint
model.constraints.add(summation(data_csvfile['Space Required'], model.x) <= 49)

opt_cbc=SolverFactory('cbc')
result = opt_cbc.solve(model)

print('\nMaximum Capacity = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print('Number of Drone',data_csvfile['Drone'][i], model.x[i].value)


Maximum Capacity =  156.0

Decision Variables
Number of Drone Sankalp 0.0
Number of Drone Kartav 0.0
Number of Drone Saahas 7.8
Number of Drone Vijay 0.0
Number of Drone Lakshya 0.0


In [38]:
#3.10 (a)MILP

model.x.domain=NonNegativeIntegers

opt_cbc=SolverFactory('cbc')
result = opt_cbc.solve(model)

print('\nMaximum Capacity = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print('Number of Drone',data_csvfile['Drone'][i], model.x[i].value)



Maximum Capacity =  155.0

Decision Variables
Number of Drone Sankalp 1.0
Number of Drone Kartav 0.0
Number of Drone Saahas 7.0
Number of Drone Vijay 0.0
Number of Drone Lakshya 0.0


In [39]:
# 3.10 (b)LP

model.x.domain=NonNegativeReals

model.constraints[5].deactivate()
model.constraints[6].deactivate()

#cost constraint
model.constraints.add(summation(data_csvfile['Cost'], model.x) <= 37)

#space constraint
model.constraints.add(summation(data_csvfile['Space Required'], model.x) <= 51)


opt_cbc=SolverFactory('cbc')
result = opt_cbc.solve(model)

print('\nMaximum Capacity = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print('Number of Drone',data_csvfile['Drone'][i], model.x[i].value)


Maximum Capacity =  148.0

Decision Variables
Number of Drone Sankalp 0.0
Number of Drone Kartav 0.0
Number of Drone Saahas 7.4
Number of Drone Vijay 0.0
Number of Drone Lakshya 0.0


In [40]:
# 3.10 (b)MILP

model.x.domain=NonNegativeIntegers

opt_cbc=SolverFactory('cbc')
result = opt_cbc.solve(model)

print('\nMaximum Capacity = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print('Number of Drone',data_csvfile['Drone'][i], model.x[i].value)


Maximum Capacity =  145.0

Decision Variables
Number of Drone Sankalp 3.0
Number of Drone Kartav 0.0
Number of Drone Saahas 5.0
Number of Drone Vijay 0.0
Number of Drone Lakshya 0.0


In [41]:
# 3.10 (c)LP

model.x.domain=NonNegativeReals

model.constraints[7].deactivate()
model.constraints[8].deactivate()

#cost constraint
model.constraints.add(summation(data_csvfile['Cost'], model.x) <= 35)

#space constraint
model.constraints.add(summation(data_csvfile['Space Required'], model.x) <= 53)


opt_cbc=SolverFactory('cbc')
result = opt_cbc.solve(model)

print('\nMaximum Capacity = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print('Number of Drone',data_csvfile['Drone'][i], model.x[i].value)


Maximum Capacity =  140.0

Decision Variables
Number of Drone Sankalp 0.0
Number of Drone Kartav 0.0
Number of Drone Saahas 7.0
Number of Drone Vijay 0.0
Number of Drone Lakshya 0.0


In [42]:
# 3.10 (c)MILP

model.x.domain=NonNegativeIntegers

opt_cbc=SolverFactory('cbc')
result = opt_cbc.solve(model)

print('\nMaximum Capacity = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print('Number of Drone',data_csvfile['Drone'][i], model.x[i].value)


Maximum Capacity =  140.0

Decision Variables
Number of Drone Sankalp 0.0
Number of Drone Kartav 0.0
Number of Drone Saahas 7.0
Number of Drone Vijay 0.0
Number of Drone Lakshya 0.0


In [43]:
# 3.10 (d)LP

model.x.domain=NonNegativeReals

model.constraints[9].deactivate()
model.constraints[10].deactivate()

#cost constraint
model.constraints.add(summation(data_csvfile['Cost'], model.x) <= 33)

#space constraint
model.constraints.add(summation(data_csvfile['Space Required'], model.x) <= 55)


opt_cbc=SolverFactory('cbc')
result = opt_cbc.solve(model)

print('\nMaximum Capacity = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print('Number of Drone',data_csvfile['Drone'][i], model.x[i].value)


Maximum Capacity =  132.0

Decision Variables
Number of Drone Sankalp 0.0
Number of Drone Kartav 0.0
Number of Drone Saahas 6.6
Number of Drone Vijay 0.0
Number of Drone Lakshya 0.0


In [44]:
# 3.10 (d)MILP

model.x.domain=NonNegativeIntegers

opt_cbc=SolverFactory('cbc')
result = opt_cbc.solve(model)

print('\nMaximum Capacity = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print('Number of Drone',data_csvfile['Drone'][i], model.x[i].value)


Maximum Capacity =  130.0

Decision Variables
Number of Drone Sankalp 2.0
Number of Drone Kartav 0.0
Number of Drone Saahas 5.0
Number of Drone Vijay 0.0
Number of Drone Lakshya 0.0


**3.10** 

On decreasing the budget 39,37,35,33 and correspondingly increasing space availability 49,51,53,55  
we see decreasing pattern in objective function values for both MILP and LP. 